gk/main.c

```diff
@@ -16,31 +16,469 @@
+			 * expires.
+			 */
+			uint64_t expire_at;
+		} declined;
+	} u;
+};
+
+/* To support the optimized implementation of generic RSS hash function. */
+static uint8_t rss_key_be[RTE_DIM(default_rss_key)];
+
+static inline uint32_t
+rss_hash_func(const void *data, __attribute__((unused)) uint32_t data_len,
+        __attribute__((unused)) uint32_t init_val)
+{
+	const struct ip_flow *flow = (const struct ip_flow *)data;
+
+	if (flow->proto == ETHER_TYPE_IPv4)
+		return rte_softrss_be((const uint32_t *)&flow->f,
+				GK_THASH_V4_L3_LEN, rss_key_be);
+	else if (flow->proto == ETHER_TYPE_IPv6)
+		return rte_softrss_be((const uint32_t *)&flow->f,
+				GK_THASH_V6_L3_LEN, rss_key_be);
+	else
+		return 0;
+}
+
+/* Type of function used to compare the hash key. */
+static int
+ip_flow_cmp_eq(const void *key1, const void *key2,
+	__attribute__((unused)) size_t key_len)
+{
+	const struct ip_flow *f1 = (const struct ip_flow *)key1;
+	const struct ip_flow *f2 = (const struct ip_flow *)key2;
+
+	if (f1->proto != f2->proto)
+		return 0;
+
+	if (f1->proto == ETHER_TYPE_IPv4)
+		return memcmp(&f1->f.v4, &f2->f.v4, sizeof(struct ipv4_addr_tuple));
+	else
+		return memcmp(&f1->f.v6, &f2->f.v6, sizeof(struct ipv6_addr_tuple));
+}
+
+/* We should avoid calling integer_log_base_2() with zero. */
+static inline uint8_t
+integer_log_base_2(uint64_t delta_time)
+{
+#if __WORDSIZE == 64
+    return (8 * sizeof(uint64_t) - 1) - __builtin_clzl(delta_time);
+#else
+    return (8 * sizeof(uint64_t) - 1) - __builtin_clzll(delta_time);
+#endif
+}
+
+/* 
+ * It converts the difference of time between the current packet and 
+ * the last seen packet into a given priority. 
+ */
+static uint8_t 
+priority_from_delta_time(uint64_t present, uint64_t past)
+{
+	uint64_t delta_time;
+
+	if (unlikely(present < past)) {
+		/*
+		 * This should never happen, but we handle it gracefully here 
+		 * in order to keep going.
+		 */
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: the present time smaller than the past time!\n");
+
+		return 0;
+	}
+
+	delta_time = (present - past) * picosec_per_cycle;
+	if (unlikely(delta_time < 1))
+		return 0;
+	
+	return integer_log_base_2(delta_time);
+}
+
+static void
+extract_packet_info(struct rte_mbuf *pkt, struct ipacket *packet)
+{
+	uint16_t ether_type;
+	struct ether_hdr *eth_hdr;
+	struct ipv4_hdr  *ip4_hdr;
+	struct ipv6_hdr  *ip6_hdr;
+
+	eth_hdr = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	ether_type = rte_be_to_cpu_16(eth_hdr->ether_type);
+
+	switch (ether_type) {
+	case ETHER_TYPE_IPv4:
+		ip4_hdr = rte_pktmbuf_mtod_offset(pkt, 
+					struct ipv4_hdr *, 
+					sizeof(struct ether_hdr));
+		packet->flow.proto = ETHER_TYPE_IPv4;
+		/*
+		 * Need to use little endian,
+		 * since the software RSS hash function takes ordering as
+		 * little endian in both bytes and bits.
+		 */
+		packet->flow.f.v4.src = rte_be_to_cpu_32(ip4_hdr->src_addr);
+		packet->flow.f.v4.dst = rte_be_to_cpu_32(ip4_hdr->dst_addr);
```

Comment:
Since the software RSS hash function is simple, I just created a new function with network order input. This can avoid these issues.